draft-aegis-mimi-arch-latest.md

```diff
@@ -1,156 +1,389 @@
+  be allowed. Unless otherwise specified, a client is a valid successor of
+  another client if their client identifiers are equal.
+
+An example interface of an identity provider is as follows:
+
+~~~
+struct {
+    HPKEPublicKey public_key;
+    Credential credential;
+} Identity;
+
+struct {
+    opaque context<V>
+} ApplicationContext;
+
+interface {
+    fn supported_credentials() -> [CredentialType];
+
+    fn validate(Identity identity, Extension group_context_ext[],
+        ApplicationContext a_ctx) -> bool;
+
+    fn client_identifier(Identity identity) -> Vec<u8>;
+
+    fn valid_successor(Identity predecessor, Identity successor) -> bool {
+        return self.client_identifier(predecessor) == 
+            self.client_identifier(successor);
+    }
+} IdentityProvider;
+~~~
+
+Each identity provider has a unique value that identifies its behavior along
+with opaque parameters that help synchronize various options across
+applications. 
+
+~~~
+uint16 IdentityProviderType;
+
+struct {
+    IdentityProviderType type;
+    opaque parameters<V>;
+} IdentityConfiguration;
+~~~
+
+## Basic Identity Provider 
+
+The basic identity provider is the minimal implementation of an identity
+provider. It has the following properties:
+
+* Basic is the only type of credential that is supported.
+* Credentials are always valid.
+* Leaf and entity identifiers are equal to the bare assertion of identity
+  provided by a BasicCredential.
+
+## X.509 Identity Provider
+
+The X.509 identify provider allows for a hierarchical identity that is based
+upon a certificate chain. x509 is the only credential type supported.
+Certificate chains MUST be validated according to the rules in RFC 5280 using
+the trust roots agreed upon by the (TODO: Some sort of extension dealing with
+trust root negotiation).
+
+~~~
+struct {
+   uint64 msg_timestamp;
+} X509ApplicationContext;
+
+struct {
+    X509IdentityRange range;
+} X509Parameters;
+~~~
+
+### X.509 Identifiers
+
+Calculating a client's identifier is done by iterating through a subset of X.509
+subject fields found within the client's certificate chain. The range of
+certificates in the chain to reference as part of the calculation is defined as
+starting with `start` values from the leaf and ending with
+`end` values from the leaf. A single unique identifier is calculated by
+hashing the Common Name value found in each certificate within the reference
+range using the current group's ciphersuite hash function.
+
+~~~
+struct {
+   uint8 start;
+   uint8 end<0..255>;
+} X509IdentityRange;
+~~~
+
+fn client_identifier(HashFunction hasher, CertificateChain cert_chain,
+X509IdentityRange range)
+{
+    for certificate in cert_chain[range.start..=range.end] {
+        let common_name = certificate.subject.common_name;
+
+        if (!common_name) {
+            throw "Common name is required";
+        }
+
+        hasher.update(common_name);
+    }
+
+    return hasher.finalize();
+}
+~~~
```

Comment:
The disadvantage of hashing is 1) looking at ID you can't tell much, e.g., what's the domain, 2) identities are different for different ciphersuites. What if the identifier is simply a list, serialized in some way? This could be easily turned into e.g. Rohan's URI scheme.

Reply:
Suggestion to make this compatible with Rohan's URI scheme:
 - Subject Names of certs in identity range can't include following reserved characters: @ / # $ :
 - Client Identifier := Subject Names of certs in identity range concatenated with : as separator between them.

This way we get:
 - Client IDs are unambiguous. E.g. client_identifier(Cert1.subjName = "AB" + Cert2.SubjName = "C") != client_identifier(Cert1.SubjName = "A" + Cert2.Subjname = "BC").
 - compatible with Rohan's URI. You can just drop in our client_identifier() output as the client/device component in a URI conforming to Rohan's scheme.